### PR TITLE
add Clover OAuth 2 provider

### PR DESCRIPTION
**Pull request by macbookandrew**

### CODE REVIEW

File: src/Clover/CloverExtendSocialite.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+use SocialiteProviders\Manager\SocialiteWasCalled;
+
+class CloverExtendSocialite
+{
+    /**
+     * Register the provider.
+     *
+     * @param \SocialiteProviders\Manager\SocialiteWasCalled $socialiteWasCalled
+     */
```
Comment (atymic) [inline src/Clover/CloverExtendSocialite.php:13]: Can remove docblock

File: src/Clover/Provider.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+{
+    /**
+     * Unique Provider Identifier.
+     */
+    const IDENTIFIER = 'CLOVER';
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopes = [''];
```
Comment (atymic) [inline src/Clover/Provider.php:20]: why empty string here? can we just leave the array empty?

File: src/Clover/Provider.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+     * Indicates if the session state should be utilized.
+     *
+     * @var bool
+     */
+    protected $stateless = true;
+
+    protected function getApiDomain(): string
+    {
+        return match (true) {
+            config('services.clover.sandbox-mode') => 'sandbox.dev.clover.com',
```
Comment (atymic) [inline src/Clover/Provider.php:32]: Use `$this->getConfig()` please

File: src/Clover/Provider.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     * {@inheritdoc}
+     */
+    protected function getTokenUrl()
+    {
+        $domain = match (true) {
+            config('services.clover.sandbox-mode') => 'apisandbox.dev.clover.com',
+            default => 'api.clover.com',
+        };
+
+        return 'https://'.$domain.'/oauth/token?'.Arr::query([
```
Comment (atymic) [inline src/Clover/Provider.php:55]: sprintf please :)

File: src/Clover/Provider.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        $requestParams = str(request()->fullUrl())
+            ->after('?')
+            ->explode('&')
+            ->mapWithKeys(fn (string $keyAndValue) => [str($keyAndValue)->before('=')->toString() => str($keyAndValue)->after('=')->toString()]);
+
+        $response = $this->getHttpClient()->get('https://'.$this->getApiDomain().'/v3/merchants/'.$requestParams['merchant_id'].'/employees/'.$requestParams['employee_id'], [
```
Comment (atymic) [inline src/Clover/Provider.php:72]: Sprintf here also please

File: src/Clover/Provider.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        $requestParams = str(request()->fullUrl())
+            ->after('?')
+            ->explode('&')
+            ->mapWithKeys(fn (string $keyAndValue) => [str($keyAndValue)->before('=')->toString() => str($keyAndValue)->after('=')->toString()]);
```
Comment (atymic) [inline src/Clover/Provider.php:70]: Huh, this is strange. The data comes back in the URL?

File: src/Clover/Provider.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        $response = $this->getHttpClient()->get('https://'.$this->getApiDomain().'/v3/merchants/'.$requestParams['merchant_id'].'/employees/'.$requestParams['employee_id'], [
+            'headers' => [
+                'Authorization' => 'Bearer '.$token,
+            ],
+        ]);
+
+        return json_decode($response->getBody(), true);
```
Comment (atymic) [inline src/Clover/Provider.php:78]:
```suggestion
        return json_decode((string) $response->getBody(), true);
```

File: src/Clover/Provider.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        ]);
+    }
+
+    /**
+     * Get the access token response for the given code.
+     *
+     * @param  string  $code
+     * @return array
+     */
+    public function getAccessTokenResponse($code)
```
Comment (atymic) [inline src/Clover/Provider.php:101]: Is this different to the parent method?

File: src/Clover/composer.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    "description": "Clover OAuth2 Provider for Laravel Socialite",
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "macbookandrew",
+            "homepage": "https://andrewrminion.com"
+        }
+    ],
+    "require": {
+        "php": "^8.0",
```
Comment (atymic) [inline src/Clover/composer.json:12]:
```suggestion
        "php": "^8.1",
```

**Review by atymic — CHANGES_REQUESTED** [review]

File: src/Clover/Provider.php (hunk 2 above)
Comment (macbookandrew) [reply src/Clover/Provider.php:20]: I think I borrowed this example from another existing provider 🤷🏻‍♂️

File: src/Clover/Provider.php (hunk 6 above)
Comment (macbookandrew) [reply src/Clover/Provider.php:70]: yeah…it’s kinda weird

File: src/Clover/Provider.php (hunk 8 above)
Comment (macbookandrew) [reply src/Clover/Provider.php]: ~yeah…Clover wants a `GET` request with just headers, no body.~ I swear I saw that somewhere and now I can’t find it 😆

**Review by atymic — APPROVED** [review]

**Comment by atymic** [comment]
Please PR to website :)

**Comment by macbookandrew** [comment]
Ugg…I just realized that due to somewhat confusing documentation, this is using OAuth v2 for part of the flow and OAuth v1 for part of it 🤦🏻‍♂️ 

And the access/refresh token returned for the OAuth v2 endpoint is not valid for use with the other API endpoints…I’m waiting on Clover support for some help on that as the documentation is kinda muddled.

I’ll submit another PR here and to the website once that’s ironed out.

**Pull request merged**